INT-4265: Add Java DSL .routeByError() EIP-method

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4265

* Add `.routeByError()` EIP-method to the `IntegrationFlowDefinition`
based on the `ErrorMessageExceptionTypeRouter`
* Add missed `dynamicChannelLimit()` option to the `RouterSpec`
* Rework `RouterSpec#RouterMappingProvider` into the `ContextRefreshedEvent`
phase initialization to be sure that dependent `MappingMessageRouterManagement`
is fully initialized before applying mapping conversions

artembilan committed:
INT-4265: Add Java DSL .routeByError() EIP-method

JIRA: https://jira.spring.io/browse/INT-4265

* Add `.routeByError()` EIP-method to the `IntegrationFlowDefinition`
based on the `ErrorMessageExceptionTypeRouter`
* Add missed `dynamicChannelLimit()` option to the `RouterSpec`
* Rework `RouterSpec#RouterMappingProvider` into the `ContextRefreshedEvent`
phase initialization to be sure that dependent `MappingMessageRouterManagement`
is fully initialized before applying mapping conversions

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/RouterSpec.java:
```diff
@@ -178,28 +198,31 @@ void addMapping(Object key, NamedComponent channel) {
-				}
-				else if (key instanceof Class) {
-					channelKey = ((Class<?>) key).getName();
+		public void onApplicationEvent(ContextRefreshedEvent event) {
```
Shouldn't we confirm that the event context is this bean's declaring context?

artembilan replied:
I think so. Will fix.
Thanks

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java:
```diff
@@ -2029,6 +2030,26 @@ public B routeToRecipients(Consumer<RecipientListRouterSpec> routerConfigurer) {
+	 * @param routerConfigurer the {@link Consumer} to provide {@link ErrorMessageExceptionTypeRouter} options.
+	 * @return the current {@link IntegrationFlowDefinition}.
+	 */
+	public B routeByError(
```
Perhaps `routeByThrowable` would be more correct?

The javadocs for `ErrorMessageExceptionTypeRouter` are incorrect too (`Exception` Vs. `Throwable`).

artembilan replied:
Well, `ErrorMessage` accepts `Throwable`, too...
We don't pursue here the strong alignment with types we use. Just human-friendly is enough, IMO...

garyrussell replied:
I agree, but specifically, here (and in the EMTR), we are routing on a `Throwable` payload, not an `ErrorMessage`.

garyrussell replied:
`Error` has a meaning in Java and we shouldn't mislead.

artembilan committed:
* Rename `routeByError()` to `routeByException()` to more reflect reality
of the `ErrorMessageExceptionTypeRouter` and don't mislead about
Java's `Error`
* Polish `ErrorMessageExceptionTypeRouter` JavaDocs a bit and some Java 8
code style
* Add `ApplicationContext` assertion into the
`RouterMappingProvider.onApplicationEvent()` do not trigger delegate
initialization if `ContextRefreshedEvent` is from the different app context

artembilan commented:
Pushed. But I renamed it into `routeByException` . Somehow I don't feel comfort with all those `.able()` in the public method API names...
If you insist, I'll rename to the `routeByThrowable()`. That is just IMHO...

garyrussell commented:
Merged as f03356ccbcdf4394e7ab3daeeb3a7362a856e5f8
